File: c_files/raylib/simple_space_survivors/src/game_manager.c
#include "../include/game_manager.h"
#include "../include/settings.h"
#include "../include/starfield.h"

#include <assert.h>
#include <math.h>

void game_manager_init(GameManager *gm, const Texture2D spritesheet)
{
    assert(gm);

    player_init(&gm->player, spritesheet);

    gm->camera = (Camera2D){
        .offset = (Vector2){GetScreenWidth() / 2.0f, GetScreenHeight() / 2.0f},
        .target = (Vector2){gm->player.dest.x, gm->player.dest.y},
        .rotation = 0.0f,
        .zoom     = 1.0f};

    starfield_init(&gm->starfield);
}

void game_manager_draw(const GameManager *gm, const Texture2D spritesheet)
{
    assert(gm);

    BeginMode2D(gm->camera);

    starfield_draw(&gm->starfield, spritesheet, gm->camera);

    player_draw(&gm->player);

    EndMode2D();
}

void game_manager_update(GameManager *gm, const float dt)
{
    assert(gm);

    player_update(&gm->player, dt, GetMousePosition());

    // update camera target
    float half_w = GetScreenWidth() / 2.0f;
    float half_h = GetScreenHeight() / 2.0f;

    gm->camera.target.x =
        fmaxf(half_w, fminf(gm->player.dest.x, MAP_SIZE - half_w));
    gm->camera.target.y =
        fmaxf(half_h, fminf(gm->player.dest.y, MAP_SIZE - half_h));
}

File: python_files/raylib/space_survivors/game_manager.py
import pyray as p

from enemy_manager import EnemyManager
from player import Player
from settings import MAP_SIZE


class GameManager:
    def __init__(self, sprite_dict: dict[str, p.Rectangle]) -> None:
        self.player: Player = Player(sprite_dict)
        self.enemy_manager: EnemyManager = EnemyManager()

        self.camera: p.Camera2D = p.Camera2D()
        self.camera.offset = p.Vector2(
            p.get_screen_width() / 2.0, p.get_screen_height() / 2.0
        )
        self.camera.target = p.Vector2(
            self.player.entity.dest.x, self.player.entity.dest.y
        )
        self.camera.rotation = 0.0
        self.camera.zoom = 1.0

    def draw_tiled_background(self, background: p.Texture) -> None:
        for y in range(0, MAP_SIZE, background.height):
            for x in range(0, MAP_SIZE, background.width):
                p.draw_texture(background, x, y, p.WHITE)

    def draw(
        self,
        background: p.Texture,
        spritesheet: p.Texture,
    ) -> None:
        p.begin_mode_2d(self.camera)

        self.draw_tiled_background(background)
        self.player.draw(spritesheet)
        self.enemy_manager.draw(spritesheet)

        p.end_mode_2d()

    def update(
        self,
        sprite_dict: dict[str, p.Rectangle],
        delta_time: float,
        mouse_pos_world: p.Vector2,
    ) -> None:

        self.player.update(sprite_dict, delta_time, mouse_pos_world)
        player_pos: p.Vector2 = p.Vector2(
            self.player.entity.dest.x, self.player.entity.dest.y
        )
        self.enemy_manager.update(sprite_dict, delta_time, player_pos)

        # update camera target
        half_w: float = p.get_screen_width() / 2.0
        half_h: float = p.get_screen_height() / 2.0

        self.camera.target.x = max(half_w, min(player_pos.x, MAP_SIZE - half_w))
        self.camera.target.y = max(half_h, min(player_pos.y, MAP_SIZE - half_h))
